codec/src/test/java/io/netty/handler/codec/ByteToMessageDecoderTest.java

```diff
@@ -562,4 +565,94 @@ protected void channelRead0(ChannelHandlerContext ctx, ByteBuf msg) throws Excep
+    }
+
+    @Test
+    public void testReuseInputBufferSufficientlyLargeToContainDuplicateMessagePartiallyReceived2x_MergeCumulator() {
+        testReusedBuffer(Unpooled.buffer(1024), true, ByteToMessageDecoder.MERGE_CUMULATOR);
+    }
+
+    @Test
+    public void testReuseInputBufferJustLargeEnoughToContainMessage_CompositeCumulator() {
+        testReusedBuffer(Unpooled.buffer(16), false, ByteToMessageDecoder.COMPOSITE_CUMULATOR);
+    }
+
+    @Test
+    public void testReuseInputBufferJustLargeEnoughToContainMessagePartiallyReceived2x_CompositeCumulator() {
+        testReusedBuffer(Unpooled.buffer(16), true, ByteToMessageDecoder.COMPOSITE_CUMULATOR);
+    }
+
+    @Test
+    public void testReuseInputBufferSufficientlyLargeToContainDuplicateMessage_CompositeCumulator() {
+        testReusedBuffer(Unpooled.buffer(1024), false, ByteToMessageDecoder.COMPOSITE_CUMULATOR);
+    }
+
+    @Test
+    public void testReuseInputBufferSufficientlyLargeToContainDuplicateMessagePartiallyReceived2x_CompositeCumulator() {
+        testReusedBuffer(Unpooled.buffer(1024), true, ByteToMessageDecoder.COMPOSITE_CUMULATOR);
+    }
+
+    private static void testReusedBuffer(ByteBuf buffer, boolean partial2, ByteToMessageDecoder.Cumulator cumulator) {
+        ByteToMessageDecoder decoder = new ByteToMessageDecoder() {
+            @Override
+            protected void decode(ChannelHandlerContext ctx, ByteBuf in, List<Object> out) {
+                while (in.readableBytes() >= 4) {
+                    int index = in.readerIndex();
+                    int len = in.readInt();
+                    assert len < (1 << 30) : "In-plausibly long message: " + len;
+                    if (in.readableBytes() >= len) {
+                        byte[] bytes = new byte[len];
+                        in.readBytes(bytes);
+                        String message = new String(bytes, CharsetUtil.UTF_8);
+                        out.add(message);
+                    } else {
+                        in.readerIndex(index);
+                        return;
+                    }
+                }
+            }
+        };
+        decoder.setCumulator(cumulator);
+        EmbeddedChannel channel = new EmbeddedChannel(decoder);
+
+        buffer.retain(); // we are pooling this buffer
+        buffer.writeInt(11); // total length of message
+        buffer.writeByte('h').writeByte('e').writeByte('l').writeByte('l');
+        if (partial2) {
+            assertFalse(channel.writeInbound(buffer)); // try reading incomplete message
+            assertTrue(channel.inboundMessages().isEmpty());
+            assertEquals(0, buffer.readerIndex(), "Incomplete message should still be readable in buffer");
+            buffer.retain(); // the same buffer is re-used in the pool
+        }
+        buffer.writeByte('o').writeByte(' ');
+        assertFalse(channel.writeInbound(buffer)); // try reading incomplete message
```

Comment:
I suggest placing the `retain` calls immediately prior to the `channel.write*` calls, so it's obvious their implied `release` is being cancelled.

Reply:
the `retain` call is currently placed where it would represent the buffer being retrieved from an allocator that decides to reuse it. I find that quite sensible. But perhaps I can improve the comment to make that clearer.

How about something like:
```
[line 632]     buffer.retain(); //  buffer is allocated from the pool, the pool would call retain()
...
[line 639/646] buffer.retain(); //  buffer is allocated from the pool - reusing same buffer, the pool would call retain()
```
Would that make things clearer?

Reply:
Yeah, the comments would make it clear what scenario the test is going for.